### ISSUE TITLE
--cache_checked_modules not seeing changes to fsti file

### ISSUE THREAD

**Issue by Chris-Hawblitzel** [FStarLang/FStar#1919]
I just saw some cases where an incremental build didn't catch some verification errors, but the full build did.  It looks like it has to do with `--cache_checked_modules`.  Suppose I have a directory with files `A.fsti`, `A.fst`, `B.fsti`, `B.fst` and nothing else (no `.checked` files yet):

```
module A
val x:int
val lemma_x:squash (x == 5)

module A
let x = 5
let lemma_x = ()

module B
open A

module B
let test () =
  assert (x == 5)
```

First, I build them all:

```
fstar.exe --cache_checked_modules --query_stats A.fsti
Verified i'face (or impl+i'face): A
All verification conditions discharged successfully

fstar.exe --cache_checked_modules --query_stats A.fst
(A.fst(3,0-3,16))       Query-stats (A.lemma_x, 1)      succeeded in 567 milliseconds ...
Verified module: A
All verification conditions discharged successfully

fstar.exe --cache_checked_modules --query_stats B.fsti
Verified i'face (or impl+i'face): B
All verification conditions discharged successfully

fstar.exe --cache_checked_modules --query_stats B.fst
(B.fst(3,2-3,17))       Query-stats (B.test, 1) succeeded in 582 milliseconds ...
Verified module: B
All verification conditions discharged successfully
```

then I modify A.fsti and A.fst:

```
module A
val x:int
//val lemma_x:squash (x == 5)

module A
let x = 6
//let lemma_x = ()
```

then I build everything again:

```
fstar.exe --cache_checked_modules --query_stats A.fsti
Verified i'face (or impl+i'face): A
All verification conditions discharged successfully

fstar.exe --cache_checked_modules --query_stats A.fst
Verified module: A
All verification conditions discharged successfully

fstar.exe --cache_checked_modules --query_stats B.fsti
Verified i'face (or impl+i'face): B
All verification conditions discharged successfully

fstar.exe --cache_checked_modules --query_stats B.fst
Verified module: B
All verification conditions discharged successfully
```

This time, `B.fst` skips verification, when it should actual fail verification.  If I delete the `.checked` files and remove `--cache_checked_modules`, then B correctly fails.

**Comment by aseemr** [comment]
Thanks @Chris-Hawblitzel for reporting the issue. I also ran into an issue with the same root cause as this bug:

```
//A.fsti:

module A
let a_predicate : prop = True

//Test.fsti
module Test
open A
val test (_:unit) : unit

//Test.fst
module Test
let test () = assert a_predicate
```

and then:

```
aseemr@ASEEMR-SBOOK ~/FStar
$ ls *.checked
ls: cannot access '*.checked': No such file or directory

aseemr@ASEEMR-SBOOK ~/FStar
$ fstar.exe Test.fst --cache_checked_modules
A.fsti(0,0-0,0): (Warning 241) Unable to load A.fsti.checked since checked file A.fsti.checked does not exist; will recheck A.fsti (suppressing this warning for further modules)
Verified module: Test
All verification conditions discharged successfully

aseemr@ASEEMR-SBOOK ~/FStar
$ ls -al Test.fst.checked
-rwxr-xr-x 1 aseemr FAREAST+Group(513) 3881 Apr  9 10:51 Test.fst.checked
```

F* should not have written `Test.fst.checked` since checked file for one of its dependency, `A.fsti`, does not exist.

The root cause of the problem is that the dependence of `Test.fst` on `A.fsti` is not direct -- it is via the `Test.fsti`. In your example as well as in mine, the `open` happens in the `.fsti` file. When F* computes the dependence graph of an `.fst`, it *DOES NOT* take into account the dependencies of its interface. So, in the dependence graph, `Test.fst` (in your example `B.fst`) does not depend on `A..fsti`. Hence the behavior of the checked files.

On the other hand, when typechecking, the `open` from `Test.fsti` is interleaved into `Test.fst`, and so, `A` is in scope.

Nik, Guido, and I discussed this. One solution we discussed it to not interleave `open` (and `include`, and module abbreviations). That way, in `Test.fst`, the programmer will explicitly have to `open A`, which will create the desired dependencies.

But then Nik observed that this will still be a problem with module abbreviations. And so, perhaps a better solution is to make `Test.fst.checked` depend on `Test.fsti.checked` (right now it only depends on `Test.fsti`). This will make `Test.fst.checked` transitively depend on the dependencies of `Test.fsti.checked`.

I have a tentative implementation of this fix which gives desired behavior with both the examples.
